tools/enrollment_preprocessor.py

```diff
@@ -0,0 +1,163 @@
+) -> dict[str, list[tuple[UserCells, str]]]:
+    groups = defaultdict(list)
+    for user_entries in user_cells.values():
+        for cells in user_entries:
+            imported = cells.clean_user()
+
+            if not imported.email:
+                continue
+
+            existing = users.setdefault(imported.email, imported)
+
+            if imported == existing:
+                continue
+
+            for field in ["title", "last_name", "first_name"]:
+                field_value = getattr(cells, field)
+                if field_value is not None and getattr(existing, field) != getattr(imported, field):
+                    groups[field].append((cells, existing.email))
+    return groups
+
+
+def parse_existing(user_data: TextIO) -> dict[str, User]:
+    users = {}
+    reader = csv.reader(user_data, delimiter=";", lineterminator="\n")
+    next(reader)  # skip header
+    for row in reader:
+        user = User(*row)
+        users[user.email] = user
+    return users
+
+
+def parse_imported(enrollment_data: Workbook):
+    user_cells: dict[str, list[UserCells]] = defaultdict(list)
+    for sheet in enrollment_data.worksheets:
+        for wb_row in sheet.iter_rows(min_row=2, min_col=2):
+            cells = UserCells(None, *wb_row[:3])
+            user_cells[cells.clean_user().email].append(cells)
+            cells = UserCells(*wb_row[7:])
+            user_cells[cells.clean_user().email].append(cells)
+    return user_cells
+
+
+def get_user_decisions(decisions: dict[str, User], import_data: dict[str, list[UserCells]]) -> dict[str, User]:
+    conflict_groups = group_conflicts(decisions, import_data)
+    for field, conflicts in conflict_groups.items():
+        print(field.capitalize())
+        print("---------")
+        for cells, existing_email in conflicts:
+            imported = cells.clean_user()
+            existing = decisions[existing_email]  # existing is current user decision
+
+            if getattr(existing, field) == getattr(imported, field):
+                continue
+
+            print(f"existing: '{make_bold(getattr(existing, field))}' ({existing.full_name()})")
+            print(f"imported: '{make_bold(getattr(imported, field))}' ({imported.full_name()})")
+
+            decision: str = ""
+            while decision not in ("e", "i"):
+                decision = input("Which one should be used? (e/i):\n")
+            choice = existing if decision == "e" else imported
+
+            setattr(decisions[choice.email], field, getattr(choice, field))
+        print()
+    return decisions
+
+
+def run_preprocessor(enrollment_data: Path | BytesIO, user_data: TextIO) -> BytesIO | None:
+    workbook = load_workbook(enrollment_data)
+
+    import_data = parse_imported(workbook)
+    decisions = get_user_decisions(
+        parse_existing(user_data),
+        import_data,
+    )
+
+    # apply decisions
+    changed = False
+    for email, user in decisions.items():
```

Comment:
Currently O(n^2). I'd just
* iterate over the rows in the import (xlsx) file
* For each row, look up the values from `decisions` (maybe call it `correct_user_data_by_email`?) and replace mismatches.


Could we just replace the cell values (and possibly overwrite with already existing data) and just check equality of the workbook we modified with the workbook we were passed to determine `changed`?

Reply:
I doubt the O(n^2) thing: In total, each cell can be in `outdated_user` at most once throughout the first loop. So, the innermost loop is executed at most the number of user cells times, which should be optimal.
Initially, I just did not want to iterate over the file again. You are right though, this might be cleaner now

Reply:
I don't think you can do that with the workbook, as the file is completely rewritten when saved. As I can see, there is no proper equality operator defined on the `Workbook` class. Also, the file changes when the identical workbook is saved again :cry: